discover/discover.go

```diff
@@ -17,77 +17,88 @@ import (
-			peek := matchedFile.Peek()
-			if peek != "" {
-				count[peek]++
-			}
+	for _, effxFileLocation := range effxFileLocations {
+		if strings.Contains(effxFileLocation, file.Name()) {
+			return true
 		}
+	}
+	return false
+}
 
-		maxK := ""
-		maxV := 1
-		for k, v := range count {
-			if v > maxV {
-				maxK = k
-				maxV = v
-			}
+func createDetectedServicePayload(file os.FileInfo, sourceName string, commonDir string) effx_api.DetectedServicesPayload {
+	payload := effx_api.DetectedServicesPayload{
+		Name:       strings.ToLower(file.Name()),
+		SourceName: &sourceName,
+		Tags:       &map[string]string{},
+	}
+
+	if os.Getenv("DISABLE_LANGUAGE_DETECTION") != "true" {
+		result, err := metadata.InferMetadata(filepath.Dir(commonDir + file.Name()))
+		if err != nil {
+			log.Printf("Could not predict version %+v\n", err)
 		}
 
-		nextRound := make([]*Iterator, 0, maxV)
-		for _, matchedFile := range matchedEffxFiles {
-			// advance ptr
-			if matchedFile.Next() == maxK {
-				// put into next
-				nextRound = append(nextRound, matchedFile)
+		if result != nil {
+			if result.Language != "" {
+				language := strings.ToLower(result.Language)
+				(*payload.Tags)["language"] = language
+
+				// version cannot be found if language cannot be detected
+				if result.Version != "" {
+					(*payload.Tags)[language] = strings.ToLower(result.Version)
+				}
 			}
 		}
+	}
 
-		prefixString += maxK
-		matchedEffxFiles = nextRound
+	return payload
+}
 
-	}
+func removeDuplicateServices(input []effx_api.DetectedServicesPayload) []effx_api.DetectedServicesPayload {
+	serviceMap := map[string]bool{}
+	result := []effx_api.DetectedServicesPayload{}
 
-	if prefixString == "" {
-		return ""
+	for _, service := range input {
+		if _, ok := serviceMap[service.Name]; !ok {
+			result = append(result, service)
+			serviceMap[service.Name] = true
+		}
 	}
 
-	// prefix string should be a directory ending with a slash
-	slashIndex := strings.LastIndex(prefixString, "/")
+	return result
+}
 
-	if slashIndex != len(prefixString) {
-		// trim file name, keep last dir slash
-		// example:
-		// services/dooku -> services/
-		prefixString = prefixString[:slashIndex+1]
+func DetectServicesFromWorkDir(workDir string, apiKeyString, sourceName string) error {
+	filePaths := parser.ProcessDirectory(workDir)
+	services, err := DetectServicesFromRelavantFiles(workDir, filePaths, sourceName)
+	if err != nil {
+		return err
 	}
 
-	return prefixString
+	servicesInferredFromYaml := DetectServicesFromEffxYamls(filePaths, apiKeyString, sourceName)
+	services = append(services, servicesInferredFromYaml...)
+	services = removeDuplicateServices(services)
+
+	return SendDetectedServices(apiKeyString, data.GenerateUrl(), services)
 }
 
-// returns a list of names of detected services
-func DetectServices(sourceName string, effxFileLocations []string) []effx_api.DetectedServicesPayload {
+// returns detected service by looking at existing effx yaml patterns
```

Comment:
`go-lint` says exported functions should have comments...

Those comments need to have the function name as the first word